infoschema/builder.go

```diff
@@ -67,13 +69,110 @@ func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) ([]int64, erro
 		return b.applyDropPolicy(diff.SchemaID), nil
 	case model.ActionAlterPlacementPolicy:
 		return b.applyAlterPolicy(m, diff)
+	default:
+		roDBInfo, ok := b.is.SchemaByID(diff.SchemaID)
+		if !ok {
+			return nil, ErrDatabaseNotExists.GenWithStackByArgs(
+				fmt.Sprintf("(Schema ID %d)", diff.SchemaID),
+			)
+		}
+		dbInfo := b.copySchemaTables(roDBInfo.Name.L)
+		switch diff.Type {
+		case model.ActionTruncateTablePartition, model.ActionTruncateTable:
+			tblIDs, err = b.applyTruncateTableOrPartition(m, diff, dbInfo)
+		case model.ActionDropTable, model.ActionDropTablePartition:
+			tblIDs, err = b.applyDropTableOrParition(m, diff, dbInfo)
+		case model.ActionRecoverTable:
+			tblIDs, err = b.applyRecoverTable(m, diff, dbInfo)
+		default:
+			tblIDs, err = b.applyDefaultAction(m, diff, dbInfo)
+		}
 	}
-	roDBInfo, ok := b.is.SchemaByID(diff.SchemaID)
-	if !ok {
-		return nil, ErrDatabaseNotExists.GenWithStackByArgs(
-			fmt.Sprintf("(Schema ID %d)", diff.SchemaID),
-		)
+	return tblIDs, err
+}
+
+func (b *Builder) applyTruncateTableOrPartition(m *meta.Meta, diff *model.SchemaDiff, dbInfo *model.DBInfo) ([]int64, error) {
+	tblIDs, err := b.applyTableUpdate(m, diff, dbInfo)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, opt := range diff.AffectedOpts {
+		if diff.Type == model.ActionTruncateTablePartition {
+			// Reduce the impact on DML when executing partition DDL. eg.
+			// While session 1 performs the DML operation associated with partition 1,
+			// the TRUNCATE operation of session 2 on partition 2 does not cause the operation of session 1 to fail.
+			tblIDs = append(tblIDs, opt.OldTableID)
+		}
+		b.applyPlacementDelete(placement.GroupID(opt.OldTableID))
+		err := b.applyPlacementUpdate(placement.GroupID(opt.TableID))
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return tblIDs, nil
+}
+
+func (b *Builder) applyDropTableOrParition(m *meta.Meta, diff *model.SchemaDiff, dbInfo *model.DBInfo) ([]int64, error) {
+	tblIDs, err := b.applyTableUpdate(m, diff, dbInfo)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, opt := range diff.AffectedOpts {
+		b.applyPlacementDelete(placement.GroupID(opt.OldTableID))
+	}
+	return tblIDs, nil
+}
+
+func (b *Builder) applyRecoverTable(m *meta.Meta, diff *model.SchemaDiff, dbInfo *model.DBInfo) ([]int64, error) {
+	tblIDs, err := b.applyTableUpdate(m, diff, dbInfo)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, opt := range diff.AffectedOpts {
+		err := b.applyPlacementUpdate(placement.GroupID(opt.TableID))
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
 	}
+	return tblIDs, nil
+}
+
+func (b *Builder) applyDefaultAction(m *meta.Meta, diff *model.SchemaDiff, dbInfo *model.DBInfo) ([]int64, error) {
+	tblIDs, err := b.applyTableUpdate(m, diff, dbInfo)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	typ := diff.Type
+	switch diff.Type {
+	case model.ActionRenameTables:
```

Comment:
It's very strange to modify to type.
I think it's better to cache `dbInfo := b.copySchemaTables(roDBInfo.Name.L)` to builder.

Reply:
> I think it's better to cache dbInfo := b.copySchemaTables(roDBInfo.Name.L) to builder.

What do you mean by "ache"? It is extracted here https://github.com/pingcap/tidb/pull/31172/files#diff-180ddcd1a841569d164350a790bb54cd07de0e7cdc48b82c595072b8ecd1eecfR73-R79. So it is reused over the whole process. 

> It's very strange to modify the type.

It is mostly to adapt this https://github.com/pingcap/tidb/pull/31172/files#diff-180ddcd1a841569d164350a790bb54cd07de0e7cdc48b82c595072b8ecd1eecfL133-R231. If you think 'RenameTables` as a batch of `RenameTable`, this makes sense. I could add comments.

Reply:
If you meant to cache `dbInfo` over the ilfetime of the builder, I think that needs more efforts than simple caching. Readonly old schema + writable new infoschema buffer will work for that.

Reply:
Removed type change. Maybe better cache and type change will be addressed in next PR.